tools/sotn-assets/build.go

```diff
@@ -0,0 +1,385 @@
+		fullPath := path.Join(path.Dir(fileName), name)
+		symbol := makeSymbolFromFileName(name)
+		eg.Go(func() error {
+			return buildGenericU16(fullPath, symbol, outputDir)
+		})
+	}
+	for name := range tiledefs {
+		fullPath := path.Join(path.Dir(fileName), name)
+		symbol := makeSymbolFromFileName(name)
+		eg.Go(func() error {
+			return buildTiledefs(fullPath, symbol, outputDir)
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+
+	layers := []map[string]interface{}{} // first layer is always empty
+	layers = append(layers, map[string]interface{}{})
+	roomLayersId := make([]int, len(roomsLayers)*2)
+	pool := map[string]int{}
+	pool[""] = 0
+	for _, rl := range roomsLayers {
+		if l, fgFound := rl["fg"]; fgFound {
+			hash := getHash(*l)
+			if index, found := pool[hash]; !found {
+				pool[hash] = len(layers)
+				roomLayersId = append(roomLayersId, len(layers))
+				layers = append(layers, pack(*l))
+			} else {
+				roomLayersId = append(roomLayersId, index)
+			}
+		} else {
+			roomLayersId = append(roomLayersId, 0)
+		}
+		if l, bgFound := rl["bg"]; bgFound {
+			hash := getHash(*l)
+			if index, found := pool[hash]; !found {
+				pool[hash] = len(layers)
+				roomLayersId = append(roomLayersId, len(layers))
+				layers = append(layers, pack(*l))
+			} else {
+				roomLayersId = append(roomLayersId, index)
+			}
+		} else {
+			roomLayersId = append(roomLayersId, 0)
+		}
+	}
+
+	sb := strings.Builder{}
+	sb.WriteString("// clang-format off\n")
+	for name := range tilemaps {
+		symbol := makeSymbolFromFileName(name)
+		sb.WriteString(fmt.Sprintf("extern unsigned short %s[];\n", symbol))
+	}
+	for name := range tiledefs {
+		symbol := makeSymbolFromFileName(name)
+		sb.WriteString(fmt.Sprintf("extern TileDefinition %s[];\n", symbol))
+	}
+
+	sb.WriteString("static MyLayer layers[] = {\n")
+	sb.WriteString("    {},\n")
+	for _, l := range layers[1:] {
+		sb.WriteString(fmt.Sprintf("    { %s, %s, 0x%08X, 0x%02X, %d, %d },\n",
+			makeSymbolFromFileName(l["data"].(string)),
+			makeSymbolFromFileName(l["tiledef"].(string)),
+			l["params"],
+			l["zPriority"],
+			l["unkE"],
+			l["unkF"],
+		))
+	}
+	sb.WriteString("};\n")
+
+	sb.WriteString("static MyRoomDef rooms[] = {\n")
+	for _, rl := range roomsLayers {
+		if l, found := rl["fg"]; found {
+			sb.WriteString(fmt.Sprintf("    { &layers[%d], ", pool[getHash(*l)]))
+		} else {
+			sb.WriteString(fmt.Sprintf("    { &layers[0], "))
+		}
+		if l, found := rl["bg"]; found {
+			sb.WriteString(fmt.Sprintf("&layers[%d] },\n", pool[getHash(*l)]))
+		} else {
+			sb.WriteString(fmt.Sprintf("&layers[0] },\n"))
+		}
+	}
+	sb.WriteString("};\n")
+	return os.WriteFile(path.Join(outputDir, "layers.h"), []byte(sb.String()), 0644)
+}
+
+func buildSpriteGroup(sb *strings.Builder, sprites [][]sprite, mainSymbol string, r *rand.Rand) {
+	symbols := []string{}
+	for _, spriteGroup := range sprites {
+		if len(spriteGroup) > 0 {
+			symbol := fmt.Sprintf("spriteGroup_%08X", r.Int31())
```

Comment:
You're not getting collisions doing this?

Reply:
I think it is impossible to get a collision as rather than doing `D_offset` I am doing `{dataType}_{randomNumber}`. If any possible collision will ever occur, it will just be a matter of randomize the symbol a bit more.